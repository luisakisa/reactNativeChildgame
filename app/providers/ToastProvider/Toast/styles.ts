import {StyleSheet} from 'react-native';
import FontStyles from '../../../constants/FontStyles';

const styles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
  },
  toast: {
    position: 'absolute',
    left: 8,
    right: 8,
    bottom: 0,
    padding: 12,
    borderRadius: 12,
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0.16,
    shadowRadius: 3.84,
    elevation: 5,
  },
  message: {
    fontWeight: '400',
    fontSize: 15,
    lineHeight: 20,
    ...FontStyles.Subheadline15Regular,
  },
});

export default styles;
