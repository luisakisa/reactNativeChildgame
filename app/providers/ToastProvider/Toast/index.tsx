import React, {memo, useContext, useEffect} from 'react';
import Animated, {
  SharedValue,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withDelay,
  withTiming,
  cancelAnimation,
  withSequence,
} from 'react-native-reanimated';
import styles from './styles';
import {LayoutChangeEvent, Text} from 'react-native';
import {TOAST_LIFETIME_DURATION} from '../config';
import {AppThemeContext} from '../../AppTheme';

const PRESENTATION_DURATION = 300;
const HIDE_DURATION = 700;

interface Props {
  message: string;
  index: number;
  toastsLength: SharedValue<number>;
}

function Toast({message, index, toastsLength}: Props): React.FunctionComponentElement<Props> {
  const theme = useContext(AppThemeContext);

  const scale = useSharedValue<number>(0.7);
  const opacity = useSharedValue<number>(0);
  const height = useSharedValue<number>(0);
  const activeToast = useSharedValue<boolean>(true);

  useEffect(() => {
    scale.value = withTiming(1, {duration: PRESENTATION_DURATION});
    opacity.value = withSequence(
      withTiming(1, {duration: PRESENTATION_DURATION}),
      withDelay(
        TOAST_LIFETIME_DURATION - PRESENTATION_DURATION - HIDE_DURATION - 100,
        withTiming(0, {duration: HIDE_DURATION}),
      ),
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const containerAnimatedStyle = useAnimatedStyle(() => {
    return {
      height: withTiming(height.value + (height.value > 0 ? 8 : 0), {
        duration: PRESENTATION_DURATION,
      }),
    };
  }, [height]);

  const toastAnimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: opacity.value,
      transform: [{scale: scale.value}],
    };
  }, [scale, opacity]);

  function handleLayout(event: LayoutChangeEvent): void {
    height.value = event.nativeEvent.layout.height;
  }

  useDerivedValue(() => {
    if (toastsLength.value !== 0 && index === toastsLength.value - 2 && activeToast.value) {
      activeToast.value = false;
      cancelAnimation(opacity);
      opacity.value = withDelay(
        PRESENTATION_DURATION + 200,
        withTiming(0, {duration: HIDE_DURATION}),
      );
    }
  }, [toastsLength]);

  return (
    <Animated.View style={[styles.container, {zIndex: 999 - index}, containerAnimatedStyle]}>
      <Animated.View
        style={[styles.toast, {backgroundColor: theme.C15}, toastAnimatedStyle]}
        onLayout={handleLayout}
      >
        <Text style={[styles.message, {color: theme.C12}]}>{message}</Text>
      </Animated.View>
    </Animated.View>
  );
}

export default memo(Toast);
