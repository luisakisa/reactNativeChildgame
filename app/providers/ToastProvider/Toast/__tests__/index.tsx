import React from 'react';
import {render} from '@testing-library/react-native';
import Toast from '../index';

test('provider/ToastProvider/Toast/snapshot', () => {
  const {toJSON} = render(<Toast index={0} message={'Toast'} toastsLength={{value: 0}} />);
  expect(toJSON()).toMatchSnapshot();
});
