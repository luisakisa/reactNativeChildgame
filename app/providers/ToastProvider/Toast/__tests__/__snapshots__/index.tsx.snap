// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider/ToastProvider/Toast/snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
      },
      Object {
        "zIndex": 999,
      },
      Object {
        "height": 0,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "borderRadius": 12,
          "bottom": 0,
          "elevation": 5,
          "left": 8,
          "padding": 12,
          "position": "absolute",
          "right": 8,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.16,
          "shadowRadius": 3.84,
        },
        Object {
          "backgroundColor": "#788CF5",
        },
        Object {
          "opacity": 0,
          "transform": Array [
            Object {
              "scale": 0.7,
            },
          ],
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 15,
            "fontWeight": "400",
            "letterSpacing": -0.24,
            "lineHeight": 20,
          },
          Object {
            "color": "#FFFFFF",
          },
        ]
      }
    >
      Toast
    </Text>
  </View>
</View>
`;
